### PR TITLE
fix: ignore `CancellationException` in suspend functions

### PR DESCRIPTION
**Pull request by Skrilltrax**
## :loudspeaker: Type of change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## :scroll: Description

Suspend functions can throw cancellation exceptions, here we are making sure that we do not catch it.

## :bulb: Motivation and Context

Had a chat with @msfjarvis on Telegram

## :green_heart: How did you test it?
All the current tests work fine, should I add a test that tries to throw `CancellationException`?

## :pencil: Checklist

- [x] I formatted the code `./gradlew spotlessApply`
- [x] I reviewed submitted code
- [ ] I added a [CHANGELOG](CHANGELOG.md) entry if applicable

### CODE REVIEW

File: coroutine-utils/src/main/kotlin/dev/msfjarvis/aps/util/coroutines/RunSuspendCatching.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import kotlin.contracts.contract
+import kotlinx.coroutines.CancellationException
+
+/**
+ * Calls the specified function [block] with [this] value as its receiver and returns its
+ * encapsulated result if invocation was successful, catching any [Throwable] except
+ * [CancellationException] that was thrown from the [block] function execution and encapsulating it
+ * as a failure.
+ */
+public suspend inline fun <V> runSuspendCatching(block: () -> V): Result<V, Throwable> {
```
Comment (msfjarvis) [inline coroutine-utils/src/main/kotlin/dev/msfjarvis/aps/util/coroutines/RunSuspendCatching.kt:20]:
```suggestion
public suspend inline fun <V> runSuspendCatching(block: suspend () -> V): Result<V, Throwable> {
```

File: coroutine-utils/src/main/kotlin/dev/msfjarvis/aps/util/coroutines/RunSuspendCatching.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+/**
+ * Calls the specified function [block] with [this] value as its receiver and returns its
+ * encapsulated result if invocation was successful, catching any [Throwable] except
+ * [CancellationException] that was thrown from the [block] function execution and encapsulating it
+ * as a failure.
+ */
+public suspend inline infix fun <T, V> T.runSuspendCatching(
+  block: T.() -> V
```
Comment (msfjarvis) [inline coroutine-utils/src/main/kotlin/dev/msfjarvis/aps/util/coroutines/RunSuspendCatching.kt:38]:
```suggestion
  block: suspend T.() -> V
```

File: coroutine-utils/build.gradle.kts [hunk 3]
```diff
@@ -10,4 +10,5 @@ plugins {
 dependencies {
   implementation(libs.kotlin.coroutines.core)
   implementation(libs.dagger.hilt.core)
+  implementation(libs.thirdparty.kotlinResult)
```
Comment (msfjarvis) [inline coroutine-utils/build.gradle.kts]:
```suggestion
  api(libs.thirdparty.kotlinResult)
```

Your public API includes types from the library so this is supposed to be an api dependency

File: coroutine-utils/src/main/kotlin/dev/msfjarvis/aps/util/coroutines/RunSuspendCatching.kt (hunk 1 above)
Comment (msfjarvis) [reply coroutine-utils/src/main/kotlin/dev/msfjarvis/aps/util/coroutines/RunSuspendCatching.kt:20]: The block should be suspend I think?

**Review by msfjarvis — CHANGES_REQUESTED** [review]

File: coroutine-utils/src/main/kotlin/dev/msfjarvis/aps/util/coroutines/RunSuspendCatching.kt (hunk 1 above)
Comment (Skrilltrax) [reply coroutine-utils/src/main/kotlin/dev/msfjarvis/aps/util/coroutines/RunSuspendCatching.kt:20]: IntelliJ threw a warning when I added `suspend` saying that a lambda parameter would use the same continuation.

```
Redundant 'suspend' modifier: lambda parameters of suspend function type uses existing continuation.
```

File: coroutine-utils/build.gradle.kts (hunk 3 above)
Comment (Skrilltrax) [reply coroutine-utils/build.gradle.kts]: Oh, missed that, my bad!

**Review by msfjarvis — APPROVED** [review]

**Pull request merged**